File: 30-day-challenge/day13/softmax.c
// write a softmax function for a matrix in c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>


float max(float a, float b){
    return (a > b) ? a : b;
}

float* softmax(float *m, float *n, int N){
    for (int v = 0; v < N; v++){
        float *a;
        a = (float *)malloc(N * sizeof(float));
        for (int i = 0; i < N; i++){
            a[i] = m[v  * 3 + i];
        }
        // find the maximu
        float m = 0;
        for (int i = 0; i < N; i++){
            m = max(a[i], m);
        }
        // find denominator values
        float s = 0;
        for (int i = 0; i < N; i++){
            s += exp(a[i] - m);
        }

        for (int i = 0; i < N; i++){
            n[v * 3 + i] = exp(a[i] - m) / s;
        }
    }
    return n;
}





int main(){
    float *h_a, *h_o;
    int N = 3;

    h_a = (float *)malloc(N * N* sizeof(float));
    h_o = (float *)malloc(N * N* sizeof(float));

    for(int i = 0; i < N; i++){
        for (int j = 0; j< N; j++){
            h_a[i * N + j] = i * N +j;
        }
    }

    for (int i = 0; i < N; i++){
        for (int j = 0; j < N; j++){
            printf("%f ", h_a[i * N + j]);
        }
    }
    printf("\n");
    h_o  = softmax(h_a, h_o, N);
    for (int i = 0; i < N; i++){
        for (int j = 0; j < N; j++){
            printf("%f ", h_o[i * N + j]);
        }
    }
    return 0;

}

File: 30-day-challenge/day13/softmax.py
import torch.nn as nn
import torch
import numpy as np
import math
import code


def torch_softmax(a):
    m = nn.Softmax(dim=0)
    output = m(a)
    print(output.numpy())


def softmax_numpy(a):
    result = np.exp(a) / sum(np.exp(a))
    print(result)


def softmax_scratch(a):
    a = a.copy()
    # find divisible factor
    s = 0
    for i in range(len(a)):
        s += math.exp(a[i])

    # device each element by division factor
    for i in range(len(a)):
        a[i] = math.exp(a[i]) / s
    print(a)


def safe_softmax_scratch(a):
    assert len(a.shape) == 1, "a can't be a matrix"
    # find maximum
    m = float('-inf')
    for i in range(len(a)):
        m = max(a[i], m)

    # find divisible factor
    s = 0
    for i in range(len(a)):
        s += math.exp(a[i] - m)

    # device each element by division factor
    for i in range(len(a)):
        a[i] = math.exp(a[i] - m) / s
    print(a)


def safe_softmax(ms):
    assert len(ms.shape) == 2, "input should be a matrix"
    for i in range(len(ms)):
        a = ms[i]

        # find maximum
        m = float('-inf')
        for i in range(len(a)):
            m = max(a[i], m)

        # find divisible factor
        s = 0
        for i in range(len(a)):
            s += math.exp(a[i] - m)

        # device each element by division factor
        for i in range(len(a)):
            a[i] = math.exp(a[i] - m) / s
    print(ms)


if __name__ == "__main__":
    a = torch.arange(3).float()
    # print('Input', a)
    # torch_softmax(a)
    # softmax_numpy(a.numpy())
    # softmax_scratch(a.numpy())
    # safe_softmax_scratch(a.numpy())

    a = torch.arange(6).reshape(2, 3).float()
    safe_softmax(a)
